core/Extendable.php

```diff
@@ -0,0 +1,465 @@
+<?php
+
+namespace SilverStripe\Framework\Core;
+
+use ClassInfo;
+use Config;
+use Extension;
+use Injector;
+
+/**
+ * Allows an object to have extensions applied to it.
+ *
+ * Bootstrap by calling $this->constructExtensions() in your class constructor.
+ *
+ * Requires CustomMethods trait
+ */
+trait Extendable {
+	use CustomMethods;
+
+	/**
+	 * An array of extension names and parameters to be applied to this object upon construction.
+	 *
+	 * Example:
+	 * <code>
+	 * private static $extensions = array (
+	 *   'Hierarchy',
+	 *   "Version('Stage', 'Live')"
+	 * );
+	 * </code>
+	 *
+	 * Use {@link Object::add_extension()} to add extensions without access to the class code,
+	 * e.g. to extend core classes.
+	 *
+	 * Extensions are instantiated together with the object and stored in {@link $extension_instances}.
+	 *
+	 * @var array $extensions
+	 * @config
+	 */
+	private static $extensions = null;
+
+	private static $classes_constructed = array();
+
+    /**
+     * Classes that cannot be extended
+     *
+     * @var array
+     */
+	private static $unextendable_classes = array('Object', 'ViewableData', 'RequestHandler');
+
+	/**
+	 * @var array all current extension instances.
+	 */
+	protected $extension_instances = array();
+
+	/**
+	 * List of callbacks to call prior to extensions having extend called on them,
+	 * each grouped by methodName.
+	 *
+	 * @var array[callable]
+	 */
+	protected $beforeExtendCallbacks = array();
+
+	/**
+	 * List of callbacks to call after extensions having extend called on them,
+	 * each grouped by methodName.
+	 *
+	 * @var array[callable]
+	 */
+	protected $afterExtendCallbacks = array();
+
+	/**
+	 * Allows user code to hook into Object::extend prior to control
+	 * being delegated to extensions. Each callback will be reset
+	 * once called.
+	 *
+	 * @param string $method The name of the method to hook into
+	 * @param callable $callback The callback to execute
+	 */
+	protected function beforeExtending($method, $callback) {
+		if(empty($this->beforeExtendCallbacks[$method])) {
+			$this->beforeExtendCallbacks[$method] = array();
+		}
+		$this->beforeExtendCallbacks[$method][] = $callback;
+	}
+
+	/**
+	 * Allows user code to hook into Object::extend after control
+	 * being delegated to extensions. Each callback will be reset
+	 * once called.
+	 *
+	 * @param string $method The name of the method to hook into
+	 * @param callable $callback The callback to execute
+	 */
+	protected function afterExtending($method, $callback) {
+		if(empty($this->afterExtendCallbacks[$method])) {
+			$this->afterExtendCallbacks[$method] = array();
+		}
+		$this->afterExtendCallbacks[$method][] = $callback;
+	}
+
+    protected function constructExtensions() {
+        $class = get_class($this);
+
+        // Register this trait as a method source
+        $this->registerExtraMethodCallback('defineExtensionMethods', function() {
```

Comment:
Yep, and if you just wanted to use ExtraMethods without extensions, you'd need to call defineMethods() instead of constructExtensions(). But yeah, it was a quick de-couple that's not perfect.